lib/std/fs/Dir.zig

```diff
@@ -1740,17 +1740,39 @@ pub fn readLinkW(self: Dir, sub_path_w: []const u16, buffer: []u8) ![]u8 {
     return std.os.windows.ReadLink(self.fd, sub_path_w, buffer);
 }
 
-/// Read all of file contents using a preallocated buffer.
-/// The returned slice has the same pointer as `buffer`. If the length matches `buffer.len`
-/// the situation is ambiguous. It could either mean that the entire file was read, and
-/// it exactly fits the buffer, or it could mean the buffer was not big enough for the
-/// entire file.
-pub fn readFile(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
+/// Read at most `buffer.len` bytes from the file at `file_path` into `buffer`.
+/// The returned slice has the same pointer as `buffer`. If the entirety of the file does not fit
+/// into `buffer`, `buffer ` contains the first `buffer.len` bytes of the file.
+pub fn readFileTruncate(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
     var file = try self.openFile(file_path, .{});
     defer file.close();
 
-    const end_index = try file.readAll(buffer);
-    return buffer[0..end_index];
+    const nb_read = try file.readAll(buffer);
+
+    return buffer[0..nb_read];
+}
+
+/// Read the entirety of the file at `file_path` into `buffer`.
+/// The returned slice has the same pointer as `buffer`. If the entirety of the file does not fit
+/// into `buffer`, `error.Truncated` is returned, and `buffer` contains the first `buffer.len` bytes
+/// of the file.
+pub fn readFile(self: std.fs.Dir, file_path: []const u8, buffer: []u8) ![]u8 {
+    var file = try self.openFile(file_path, .{});
+    defer file.close();
+
+    var unused: [1]u8 = undefined;
+    var iovecs = [_]std.os.iovec{
+        .{ .iov_base = buffer.ptr, .iov_len = buffer.len },
+        .{ .iov_base = &unused, .iov_len = unused.len },
+    };
+
+    const nb_read = try file.readvAll(&iovecs);
+
+    if (nb_read > buffer.len) {
```

Comment:
> When used on a pipe, this causes data loss

Please demonstrate a case where the file is immediately opened, then closed, and data loss is important.

> im not convinced the fix is an improvement over the bug

Then please suggest a better fix.

Reply:
I appreciate your commitment to ensuring the stdlib works well even in edge cases, but it would have been nice to test this yourself first. In my tests with a FIFO and the `readv`, and truncating variants, the writing process always gets `SIGPIPE`'d as soon as the reading process closes the file, data loss is not a concern. In the common/intended case, where the FIFO contains less than `buffer.len` bytes, the implementations are indistinguishable to the reader:

<details> <summary>Implementations</summary>

`read_file_truncate.zig`
```zig
const std = @import("std");

const Dir = std.fs.Dir;

pub fn readFileTruncate(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
    var file = try self.openFile(file_path, .{});
    defer file.close();

    const nb_read = try file.readAll(buffer);

    return buffer[0..nb_read];
}

pub fn main() !void {
    var buffer: [6969]u8 = undefined;
    _ = try readFileTruncate(std.fs.cwd(), "myfifo", &buffer);
}
```

`read_file_readv.zig`
```zig
const std = @import("std");

const Dir = std.fs.Dir;

pub fn readFileReadv(self: std.fs.Dir, file_path: []const u8, buffer: []u8) ![]u8 {
    var file = try self.openFile(file_path, .{});
    defer file.close();

    var unused: [1]u8 = undefined;
    var iovecs = [_]std.os.iovec{
        .{ .iov_base = buffer.ptr, .iov_len = buffer.len },
        .{ .iov_base = &unused, .iov_len = unused.len },
    };

    const nb_read = try file.readvAll(&iovecs);

    if (nb_read > buffer.len) {
        return error.Truncated;
    }

    return buffer[0..nb_read];
}

pub fn main() !void {
    var buffer: [6969]u8 = undefined;
    _ = try readFileReadv(std.fs.cwd(), "myfifo", &buffer);
}
```

</details>

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "00000000: 0000 0000 0000 0000 00"..., 6969) = 68
read(3, "", 6901)                       = 0
close(3)                                = 0
+++ exited with 0 +++
```
```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="00000000: 0000 0000 0000 0000 00"..., iov_len=6969}, {iov_base="", iov_len=1}], 2) = 68
readv(3, [{iov_base="", iov_len=6901}, {iov_base="", iov_len=1}], 2) = 0
close(3)                                = 0
+++ exited with 0 +++
```

In the intended but unlikely case where the file has exactly `buffer.len` bytes, the `readv` variant uses 4 syscalls: 

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., 6969) = 6969
close(3)                                = 0
+++ exited with 0 +++
```

```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., iov_len=6969}, {iov_base="", iov_len=1}], 2) = 6969
readv(3, [{iov_base="", iov_len=1}], 1) = 0
close(3)                                = 0
+++ exited with 0 +++
```
In the unintended case where the file has more than `buffer.len` bytes, the truncating variant behaves incorrectly in 3 syscalls:

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "00000000: 0000 0000 0000 0000 00"..., 6969) = 6969
close(3)                                = 0
+++ exited with 0 +++
```

The `readv` variant behaves correctly, reading one more byte than the buffer in 3 syscalls:

```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="00000000: 0000 0000 0000 0000 00"..., iov_len=6969}, {iov_base="0", iov_len=1}], 2) = 6970
close(3)                                = 0
error: Truncated
(..., stacktrace follows)
```